File: sorting/java/src/ShellSort.java
import java.util.*;
import java.io.File;

public class ShellSort extends SortingAlgorithm {

	// default value of increment sequence from the shellsort paper
	int[] sequence = {1391376, 463792, 198768, 86961, 33936, 13776, 4592, 1968, 861, 336, 112, 48, 21, 7, 3, 1};

	public ShellSort( int[] arr, int n)
	{
		super( arr, n ); 
	}

	public void sort()
	{

		for(int k=0; k < this.sequence.length; k++) 
		{
			int h = this.sequence[k];

			for(int i=0; i < n; i++)
			{
				int index = i + h;
				
				// check if the current index is already outside the array
				// break the current loop if it is
				// this will also ignore increment sequence values 
				// that is outside the array bounds
				if (index >= n)	break;

				int current = arr[index];
				while (index-h >= 0 && current < arr[index -h])
				{
					arr[index] = arr[index - h];
					index -= h;
				}
				arr[index] = current;
			}
		}
	}

	public void setSequence(int[] seq)
	{
		this.sequence = seq;
	}
	
}

File: sorting/python/shellsort.py
import unittest


def shell_sort(A):
    seq = [4, 2, 1]
    # h = 4
    for k in range(len(seq)):
        h = seq[k]
        for i in range(0, len(A)):
            index = i + h

            if index >= len(A):
                break

            current = A[index]
            while index-h >= 0 and current < A[index - h]:
                A[index] = A[index - h]
                index -= h
            A[index] = current  # index because index-1 defeats current
    return A


class UnitTestClass(unittest.TestCase):

    def test_shell_sort(self):
        self.assertEqual(shell_sort([12, 3, 7, 8, 5, 9, 13, 17, 2, 8, 1]), [
                         1, 2, 3, 5, 7, 8, 8, 9, 12, 13, 17])

if __name__ == "__main__":
    unittest.main()
